### PR TITLE
Add async image manage

### PR DESCRIPTION
**Pull request by taltaub22**

### CODE REVIEW

File: logic/imagesLogic.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  })
+}
+
+function deleteImage (name) {
+  return fs.unlink(path.resolve(IMAGES_DIR, name))
+}
+
+function createReturnObject (imageName) {
+  return new Promise((resolve) => {
+    base64(path.resolve(IMAGES_DIR, imageName)).then(data => {
```
Comment (idanstark42) [inline logic/imagesLogic.js]: No errors can be thrown here? What if the file suddenly doesn't exist, or we don't have "read" permissions. Always catch in IO operation

File: index.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ const {loggerMiddleware, Logger} = require('@first-lego-league/ms-logger')
 
 const logger = Logger()
 const crudRouter = require('./routers/crudRouter').getRouter
+const { initImagesFolder } = require('./logic/imagesLogic')
```
Comment (idanstark42) [inline index.js]: Seems like the index should not be familiar with the logic... I think this initiation, beside it needing to be async, shuold accour in the router

File: package.json [hunk 3]
```diff
@@ -15,12 +15,15 @@
...
+    "base64-img-promise": "^1.0.5",
     "body-parser": "^1.18.3",
     "cors": "^2.8.4",
     "express": "^4.16.3",
+    "fs-extra": "^7.0.0",
     "mhub": "^0.9.1",
     "moment": "^2.22.1",
-    "mongodb": "^3.1.0-beta4"
+    "mongodb": "^3.1.0-beta4",
+    "promisify": "^0.0.3"
```
Comment (idanstark42) [inline package.json]: You don't use this....

File: routers/imagesRouter.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  getImage(req.params.imageName).then(image => {
+    res.send(image)
+  }).catch(err => {
+    res.status(500)
+    res.send(err.message)
+  })
+})
+
+router.post('/', adminAction, (req, res) => {
+  saveImageFromBase64(req.body.name, req.body.image).then(() => {
```
Comment (idanstark42) [inline routers/imagesRouter.js]: The client has to send you a base64 of the image? You already have a package to do this yourself, why can't the cient simply give you a file?

Comment (taltaub22) [reply routers/imagesRouter.js]: I think its easier to send text, what do you think?

File: logic/imagesLogic.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+const MsLogger = require('@first-lego-league/ms-logger').Logger()
+const path = require('path')
+const fs = require('fs-extra')
+const {base64, img} = require('base64-img-promise')
+
+const IMAGES_DIR = path.resolve(process.env.DATA_DIR, 'images')
+const ALLOWED_FORMATS = ['jpg', 'jpeg', 'png', 'gif']
+
+function initImagesFolder () {
```
[the remark this answers is not in the document]
Comment (taltaub22) [reply logic/imagesLogic.js]: This method occur before the server loads, and it makes sure we have an image folder

File: routers/imagesRouter.js (hunk 4 above)
Comment (idanstark42) [reply routers/imagesRouter.js]: From the client? I dont see the difference between sending a file and a text, as long as it's not part of a larger form, which it isn't. The cilent shouldn't need to worry about the base64 operations if it doesn't have to

File: logic/imagesLogic.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (idanstark42) [reply logic/imagesLogic.js]: It still doesn't need to be sync. It blocks the loading of the server. This means that if the images mechanism doesn't work, the entire tournament service wouldn't work

File: routers/imagesRouter.js (hunk 4 above)
Comment (taltaub22) [reply routers/imagesRouter.js]: The difference is the parsing on the server side. If we define that we get it as a text, I dont need special handling of form data and i can get it as json

**Comment by alanggreen** [comment]
Would it not be a good idea to write an error message to the logfile in the
.catch in imagesRouter.js?
I'm thinking of problems out in the field that will be hard to follow if
the log doesn't say anything.
Also, in general, sending the err.message to the end user could sometimes
(or maybe even often) not look good.

**Comment by taltaub22** [comment]
@alanggreen  The client need to handle an error response. 
And yes, you are correct, Adding it now.

File: routers/imagesRouter.js (hunk 4 above)
Comment (idanstark42) [reply routers/imagesRouter.js]: This should not be a worry for the client. The client is now bound to the Base64 implementation that the server chose. The image has to be a file, but ot doesn't have to be in Base64. You should stick to what's constant.

About the "special handling":
This is [formidable](https://www.npmjs.com/package/formidable), which does the work for you, and here is a [W3Schools guid](https://www.w3schools.com/nodejs/nodejs_uploadfiles.asp) on how to use it.
